### PR TITLE
Removed UnnecessaryAbstractClass if it inherits from a abstract class

### PR DESCRIPTION
**Pull request by gouri-panda**
closes: #4753 

Removed UnnecessaryAbstractClass if it inherits from abstract class an interface or that has concrete members.

### CODE REVIEW

File: detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryAbstractClassSpec.kt [hunk 1]
```diff
@@ -73,8 +73,7 @@ class UnnecessaryAbstractClassSpec(val env: KotlinCoreEnvironment) {
                     }
                     abstract class B : A
                 """
-                val findings = subject.compileAndLintWithContext(env, code)
-                assertFindingMessage(findings, message)
+                assertThat(subject.compileAndLintWithContext(env, code)).isEmpty()
```
Comment (BraisGabin) [inline detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryAbstractClassSpec.kt:76]: This test shouldn't change, right?

File: detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryAbstractClassSpec.kt [hunk 2]
```diff
@@ -86,6 +85,16 @@ class UnnecessaryAbstractClassSpec(val env: KotlinCoreEnvironment) {
                     }
                     abstract class B : A()
                 """
+                assertThat(subject.compileAndLintWithContext(env, code)).isEmpty()
+            }
+
+            @Test
+            fun `that inherits from another empty abstract class`() {
+                val code = """
+                    @Deprecated("We don't care about this first class")
```
Comment (schalkms) [inline detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryAbstractClassSpec.kt:94]: I think this annotation can be removed.

```suggestion
```

**Review by schalkms — APPROVED** [review]
Thanks for the fix - appreciate it!

File: detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryAbstractClassSpec.kt (hunk 1 above)
Comment (gouri-panda) [reply detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryAbstractClassSpec.kt:76]: @BraisGabin If we don't change it, the tests will fail.  I don't understand your question.

File: detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryAbstractClassSpec.kt (hunk 2 above)
Comment (gouri-panda) [reply detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryAbstractClassSpec.kt:94]: @schalkms Without this, it fails.

File: detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryAbstractClassSpec.kt (hunk 1 above)
Comment (BraisGabin) [reply detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryAbstractClassSpec.kt:76]: But this one was right. An abstract class that only extends an interface is an unnecessary abstract class. The only one that should change is the case where an abstract class extend another abstract class because there you can't use an interface. (maybe I'm missing something here)

Comment (gouri-panda) [reply detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryAbstractClassSpec.kt:76]: Should I revert that?

Comment (BraisGabin) [reply detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryAbstractClassSpec.kt:76]: Yes. This test shouldn't change.

File: detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryAbstractClass.kt [hunk 3]
```diff
@@ -98,6 +98,7 @@ class UnnecessaryAbstractClass(config: Config = Config.empty) : Rule(config) {
                     report(CodeSmell(issue, Entity.from(nameIdentifier), noConcreteMember))
                 }
             }
+            hasInheritedMember(true) -> return
```
Comment (amitdash291) [inline detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryAbstractClass.kt:101]: The following enhancement to this logic will ensure the rule still fails for `interface` super types as mentioned in the earlier comment related to the `that inherits from an interface` test case.

```
hasInheritedMember(true) && !isParentInterface() -> return

private fun KtClass.isParentInterface() =
    (bindingContext[BindingContext.CLASS, this]?.unsubstitutedMemberScope as? LazyClassMemberScope)
        ?.supertypes?.firstOrNull()?.isInterface() == true
```

Comment (gouri-panda) [reply detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryAbstractClass.kt]: @amitdash291 Thanks for advice and code :)

File: detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryAbstractClassSpec.kt [hunk 4]
```diff
@@ -86,6 +86,16 @@ class UnnecessaryAbstractClassSpec(val env: KotlinCoreEnvironment) {
...
+                    abstract class A 
+                    abstract class B : A()
+                """
```
Comment (BraisGabin) [inline detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryAbstractClassSpec.kt:98]: I'm not sure, why is this raises a finding? Isn't this nearly the same case as the one before?

Comment (gouri-panda) [reply detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryAbstractClassSpec.kt:98]: Sure! I'll remove this.

File: detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryAbstractClass.kt (hunk 3 above)
Comment (amitdash291) [reply detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryAbstractClass.kt]: I think there's a corner case to test here. What if the class implements an interface and inherits from an Abstract class in that order. For eg may fail the rule whereas it's expected to pass:
```
interface I

abstract class A
// assume some fields for class A

abstract class B: I, A()
```

**Review by amitdash291 — CHANGES_REQUESTED** [review]
Refer reply to the code suggestion earlier.

File: detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryAbstractClass.kt (hunk 3 above)
Comment (BraisGabin) [reply detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryAbstractClass.kt]: Is that even valid Kotlin code? Shouldn't the class be defined the first one? I can't check right now. Anyway, a test extending both, a class and an interface seems like a really good case to have a test. @guori-panda, can you add such test?

Comment (amitdash291) [reply detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryAbstractClass.kt]: @BraisGabin yes interestingly this is valid Kotlin code. A class can implement multiple interfaces and inherits from a class at the same time.

**Comment by gouri-panda** [comment]
@BraisGabin @amitdash291 Sure :)

File: detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryAbstractClassSpec.kt [hunk 5]
```diff
@@ -88,6 +88,21 @@ class UnnecessaryAbstractClassSpec(val env: KotlinCoreEnvironment) {
...
+            @Test
+            fun `does not report abstract class that inherits from an interface and an abstract class in that order`() {
+                val code = """
+                interface I
+
+                @Deprecated("We don't care about this first class")
+                abstract class A {
+                    abstract val i: Int
+                }
+                abstract class B: A(), I
```
Comment (BraisGabin) [inline detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryAbstractClassSpec.kt:101]:
```suggestion
                abstract class B: A(), I
                abstract class C: I, A()
```

We should check both cases.

**Review by BraisGabin — COMMENTED** [review]
The PR still has conflicts.

File: detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryAbstractClassSpec.kt (hunk 5 above)
Comment (gouri-panda) [reply detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryAbstractClassSpec.kt:101]: Sure! Let me add another test case as well!

Comment (amitdash291) [reply detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryAbstractClassSpec.kt:101]: @gouri-panda I have an idea to solve this in a generic manner. If you're okay I can raise a PR on your fork branch `gouri-panda:fix-4753`?

Comment (gouri-panda) [reply detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryAbstractClassSpec.kt:101]: I was thinking if we check one of the parent is abstract, then we can return. But go for it :)

Comment (amitdash291) [reply detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryAbstractClassSpec.kt:101]: @gouri-panda my idea was to check if all parents are interfaces, basically the complementary set of whether any parent is abstract. The reason being I'm unable to find any helper method indicating if the parent class is abstract.

Comment (amitdash291) [reply detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryAbstractClassSpec.kt:101]: @gouri-panda 
I have pushed my changes as a [PR to your fork's branch here](https://github.com/gouri-panda/detekt/pull/1).
In summary I have checked if not all the parents are interfaces, which is equivalent of the fact that at least one parent is abstract.
The [commit is here](https://github.com/gouri-panda/detekt/pull/1/commits/69b3fd32157890e824e84781471fc660a14a19fd) (there are a few other merge commits)

**Review by BraisGabin — APPROVED** [review]

**Comment by BraisGabin** [comment]
Thanks for the fix @gouri-panda and @amitdash291
